### DISCUSSION TITLE
How to approach "private" methods in declarative API interfaces?

### DISCUSSION THREAD

**Discussion by natan-abolafya** [jdbi/jdbi#2668] [General]
In case what declarative api is isn't clear: https://jdbi.org/#_declarative_api

So I have quite a few of these for some more complex setups:

```
public interface UserDao {
    default List<User> list() {
         var users = listUsers();
         // do something with the users here, maybe combine with some other tables etc.
         return users;
    }

    @SqlQuery("SELECT * FROM \"user\" ORDER BY \"name\"")
    @RegisterBeanMapper(User.class)
    List<User> listUsersSql();
}
```

So, normally, I use the `list()` in the logic part of the code. But it happened a couple times that I made the mistake of calling `listUsersSql()` instead which kind of works until an obscure bug appears later because the relevant relation isn't there. Definitely my fault with the initial method naming really.

So, I want to address that by renaming the `listUsersSql` and friends to something less approachable as I can't set them `private` in an interface. But I can't decide what kind of convention to use. Any ideas there?
~~Or (🦆 effect) should I change those interfaces to classes and go for `protected abstract`?~~ I forgot that JDBI 3 dropped supporting abstract classes.

**Comment by stevenschlansker** [comment]
Yes, this is a bit of a limitation of not supporting abstract class DAOs.
Actually, it is a bit more supported now, if you are interested in trying out a beta-quality feature: https://jdbi.org/#generator

Another way to factor this with pure interfaces is to separate the high level and low level operations:

```
class UserStore {
    List<User> list() { return doStuffWith(userStoreDao.list()); }
}

interface UserStoreDao {
    List<User> listUsersSql();
}
```

It is a little bit more code but strongly divides the public interfaces from internal ops.
